RTOS.c

```diff
@@ -185,39 +251,53 @@ void waitOnSemaphor(semaphore_t *sem){
 	else{
 		//semaphore is closed, wait until it is signalled
 		runningTCB->state = WAITING;
-		addToList(runningTCB, &(sem->waitListHead));
+		addToList(runningTCB, sem->waitingPriorityQueue);
 		forceContextSwitch();
 	}
 	__enable_irq();
 }
 
-void signalSemaphor(semaphore_t *sem){
+void signalSemaphore(semaphore_t *sem){
 	__disable_irq();
 	sem->count++;
-	if(sem->waitListHead != NULL){
-		TCB_t *temp = sem->waitListHead->next;
-		sem->waitListHead->next = NULL;
-		sem->waitListHead->state = READY;
-		addToList(sem->waitListHead, &readyListHead);
-		sem->waitListHead = temp;
+
+	//check if there is a task waiting for semaphore
+  for (taskPriority_t priority = HIGHEST_PRIORITY; priority < NUM_PRIORITIES; priority++){
+		if (sem->waitingPriorityQueue[priority]->head!=NULL){
+			TCB_t *unblockedTask = sem->waitingPriorityQueue[priority]->head;
+
+			if(sem->waitingPriorityQueue[priority]->head->next != NULL){ //if not only task in list
```

Comment:
should probably change `sem->waitingPriorityQueue[priority]->head` to `unblockedTask`